meta-facebook/op2-op/src/platform/plat_power_seq.c

```diff
@@ -0,0 +1,783 @@
+		.p3v3_efuse_power_good = OPB_PWRGD_P3V3_E1S_2_R,
+		.clkbuf_oe_en = OPB_CLKBUF_E1S_2_OE_N,
+		.pcie_reset = OPB_RST_E1S_2_PERST },
+	[3] = { .present = OPB_E1S_3_PRSNT_N,
+		.p12v_efuse_enable = OPB_P12V_E1S_3_EN_R,
+		.p12v_efuse_power_good = OPB_PWRGD_P12V_E1S_3_R,
+		.p3v3_efuse_enable = OPB_P3V3_E1S_3_EN_R,
+		.p3v3_efuse_power_good = OPB_PWRGD_P3V3_E1S_3_R,
+		.clkbuf_oe_en = OPB_CLKBUF_E1S_3_OE_N,
+		.pcie_reset = OPB_RST_E1S_3_PERST },
+	[4] = { .present = OPB_E1S_4_PRSNT_N,
+		.p12v_efuse_enable = OPB_P12V_E1S_4_EN_R,
+		.p12v_efuse_power_good = OPB_PWRGD_P12V_E1S_4_R,
+		.p3v3_efuse_enable = OPB_P3V3_E1S_4_EN_R,
+		.p3v3_efuse_power_good = OPB_PWRGD_P3V3_E1S_4_R,
+		.clkbuf_oe_en = OPB_CLKBUF_E1S_4_OE_N,
+		.pcie_reset = OPB_RST_E1S_4_PERST },
+};
+
+uint8_t get_e1s_present(uint8_t device_index)
+{
+	uint8_t card_type = get_card_type();
+	uint8_t present = 0;
+
+	switch (card_type) {
+	case CARD_TYPE_OPA:
+		present = gpio_get(opa_e1s_power_control_gpio[device_index].present);
+		break;
+	case CARD_TYPE_OPB:
+		present = gpio_get(opb_e1s_power_control_gpio[device_index].present);
+		break;
+	default:
+		LOG_ERR("UNKNOWN CARD TYPE");
+		break;
+	}
+	return present;
+}
+
+uint8_t get_e1s_power_good(uint8_t device_index)
+{
+	uint8_t card_type = get_card_type();
+	uint8_t power_good = 0;
+
+	switch (card_type) {
+	case CARD_TYPE_OPA:
+		power_good =
+			(gpio_get(opa_e1s_power_control_gpio[device_index].p12v_efuse_power_good) &
+			 gpio_get(opa_e1s_power_control_gpio[device_index].p3v3_efuse_power_good));
+		break;
+	case CARD_TYPE_OPB:
+		power_good =
+			(gpio_get(opb_e1s_power_control_gpio[device_index].p12v_efuse_power_good) &
+			 gpio_get(opb_e1s_power_control_gpio[device_index].p3v3_efuse_power_good));
+		break;
+	default:
+		LOG_ERR("UNKNOWN CARD TYPE");
+		break;
+	}
+	return power_good;
+}
+
+void control_power_stage(uint8_t control_mode, uint8_t control_seq)
+{
+	switch (control_mode) {
+	case ENABLE_POWER_MODE: // Control power on stage
+	case HIGH_DISABLE_POWER_MODE:
+		if (gpio_get(control_seq) != POWER_ON) {
+			gpio_set(control_seq, POWER_ON);
+		}
+		break;
+	case LOW_ENABLE_POWER_MODE:
+	case DISABLE_POWER_MODE: // Control power off stage
+		if (gpio_get(control_seq) != POWER_OFF) {
+			gpio_set(control_seq, POWER_OFF);
+		}
+		break;
+	default:
+		LOG_ERR("Not support control mode 0x%x", control_mode);
+		break;
+	}
+}
+
+int check_power_stage(uint8_t check_mode, uint8_t check_seq)
+{
+	int ret = 0;
+	switch (check_mode) {
+	case ENABLE_POWER_MODE: // Control power on stage
+	case HIGH_DISABLE_POWER_MODE:
+		if (gpio_get(check_seq) != POWER_ON) {
+			ret = -1;
+		}
+		break;
+	case LOW_ENABLE_POWER_MODE:
+	case DISABLE_POWER_MODE: // Control power off stage
+		if (gpio_get(check_seq) != POWER_OFF) {
+			ret = -1;
+		}
+		break;
+	default:
+		LOG_ERR("Check mode 0x%x not supported!", check_mode);
+		ret = 1;
+		break;
+	}
+
+	if (ret == -1) { // Addsel if check power stage fail
+		//TBD

```

Comment:
If this is an error condition, at a minimum it should log it as such.

Reply:
Modify as recommendation.